packages/x-charts-pro/src/FunnelChart/curves/linear.ts

```diff
@@ -26,102 +29,110 @@ const yFromX =
+    this.points.push({ x: xIn, y: yIn });
+    if (this.points.length < 4) {
+      return;
     }
-    this.line = 1 - this.line;
-  }
 
-  point(x: number, y: number): void {
-    x = +x;
-    y = +y;
-
-    // We draw the lines only at currentPoint 1 & 3 because we need
-    // The data of a pair of points to draw the lines.
-    // Hence currentPoint 1 draws a line from point 0 to point 1 and point 1 to point 2.
-    // currentPoint 3 draws a line from point 2 to point 3 and point 3 to point 0.
-
-    if (this.isHorizontal) {
-      const yGetter = yFromX(this.x, this.y, x, y);
-      let xGap = 0;
-
-      // 0 is the top-left corner.
-      if (this.currentPoint === 1) {
-        xGap = this.x + this.gap;
-        this.context.moveTo(xGap, yGetter(xGap));
-        this.context.lineTo(xGap, yGetter(xGap));
-        xGap = x - this.gap;
-        this.context.lineTo(xGap, yGetter(xGap));
-      } else if (this.currentPoint === 3) {
-        xGap = this.x - this.gap;
-        this.context.lineTo(xGap, yGetter(xGap));
-        xGap = x + this.gap;
-        this.context.lineTo(xGap, yGetter(xGap));
+    // Add gaps where they are needed.
+    this.points = this.points.map((point, index) => {
+      const slopeStart = this.points.at(index <= 1 ? 0 : 2)!;
+      const slopeEnd = this.points.at(index <= 1 ? 1 : 3)!;
+      const yGetter = yFromX(
+        slopeStart.x - this.gap,
+        slopeStart.y,
+        slopeEnd.x - this.gap,
+        slopeEnd.y,
+      );
+      if (this.isHorizontal) {
+        const xGap = point.x + (index === 0 || index === 3 ? this.gap : -this.gap);
+
+        return {
+          x: xGap,
+          y: yGetter(xGap),
+        };
       }
-    }
 
-    if (!this.isHorizontal) {
-      const xGetter = xFromY(this.x, this.y, x, y);
-      let yGap = 0;
-
-      // 0 is the top-right corner.
-      if (this.currentPoint === 1) {
-        yGap = this.y + this.gap;
-        this.context.moveTo(xGetter(yGap), yGap);
-        this.context.lineTo(xGetter(yGap), yGap);
-        yGap = y - this.gap;
-        this.context.lineTo(xGetter(yGap), yGap);
-      } else if (this.currentPoint === 3) {
-        yGap = this.y - this.gap;
-        this.context.lineTo(xGetter(yGap), yGap);
-        yGap = y + this.gap;
-        this.context.lineTo(xGetter(yGap), yGap);
+      const xGetter = xFromY(
+        slopeStart.x,
+        slopeStart.y - this.gap,
+        slopeEnd.x,
+        slopeEnd.y - this.gap,
+      );
+      const yGap = point.y + (index === 0 || index === 3 ? this.gap : -this.gap);
+      return {
+        x: xGetter(yGap),
+        y: yGap,
+      };
+    });
+
+    const getBorderRadius = () => {
+      if (this.gap > 0) {
+        return this.borderRadius;
       }
-    }
+      if (this.position === 0) {
+        return [0, 0, this.borderRadius, this.borderRadius];
+      }
+      if (this.position === this.sections - 1) {
+        return [this.borderRadius, this.borderRadius];
+      }
+      return 0;
+    };
```

Comment:
Should we move this to a private class method? It doesn't reference any variable in scope.